relayer/crates/starknet-integration-tests/src/tests/ics20.rs

```diff
@@ -330,188 +360,239 @@ fn test_starknet_ics20_contract() -> Result<(), Error> {
         info!("starknet_channel_id: {:?}", starknet_channel_id);
         info!("cosmos_channel_id: {:?}", cosmos_channel_id);
 
-        // submit dummy PacketRecv message to IBC core contract
+        // submit ics20 transfer to Cosmos
 
-        // stub
-        let sender_address = "cosmos1wxeyh7zgn4tctjzs0vtqpc6p5cxq5t2muzl7ng".to_string();
+        let wallet_cosmos_a = &cosmos_chain_driver.user_wallet_a;
+        let address_cosmos_a = &wallet_cosmos_a.address;
+        let wallet_starknet_b = &starknet_chain_driver.user_wallet_b;
+        let address_starknet_b = &wallet_starknet_b.account_address;
+        let transfer_quantity = 1000;
+        let denom_cosmos = &cosmos_chain_driver.genesis_config.transfer_denom;
 
-        let recipient_address = starknet_chain_driver.user_wallet_a.account_address;
+        let balance_cosmos_a_step_0 = cosmos_chain
+            .query_balance(address_cosmos_a, denom_cosmos)
+            .await?;
 
-        let amount = 99u32;
+        info!(
+            "cosmos balance before transfer: {:?}",
+            balance_cosmos_a_step_0
+        );
 
-        let starknet_client_state = {
-            starknet_chain
-                .query_client_state(
-                    PhantomData::<CosmosChain>,
-                    &starknet_client_id,
-                    &starknet_chain.query_chain_height().await?,
+        let packet = <CosmosChain as CanIbcTransferToken<StarknetChain>>::ibc_transfer_token(
+            cosmos_chain,
+            &cosmos_channel_id,
+            &IbcPortId::transfer(),
+            wallet_cosmos_a,
+            address_starknet_b,
+            &Amount::new(transfer_quantity, denom_cosmos.clone()),
+            &None,
+        )
+        .await?;
+
+        // TODO(rano): how do I get the ics20 token contract address from starknet events
```

Comment:
When a token is created, the ICS20 emits a `CreateTokenEvent` that includes the token's address. We can read the address by retrieving that event.

Reply:
The comment is mainly about how to use `relay_packet` and still get the token address. The hack, that you pointed in https://github.com/informalsystems/ibc-starknet/pull/172#discussion_r1907728623, is introduced exactly for this.